Upgrade to newer versions of MSVC compiler

**Build Error Description**
As reported in #389, Visual Studio after 2015 Update 1 will not compile BitShares. As time marches on, acquiring and compiling with older versions becomes difficult (and in the future unsupported by their vendor). Some time should be spent exploring the issues with these newer compilers.

The first issue that arises seems to be associated with differing parameter types in a functor (see output log below). My initial triage seems to indicate that the database_api functions that have parameters (string, string) work fine. Functions that have parameters (string, string, unsigned) fail. Removing the unsigned parameter gets us further (thereby indicating it was the unsigned that was the issue), but of course is not a viable solution.

My guess is that Microsoft is being (overly?) picky about parameter types, or our templating-heavy API functor conversion is specified incorrectly in the eyes of MSVC. 

**Build Environment**
 - Host OS:             Windows 10
 - Host Physical RAM    8GB
 - Source Branch/Tag:   develop
 - OpenSSL Version:     1.1.0f
 - Boost Version:       1.65.1 (1.69.1 works better)
 - C++ Compiler:       Visual Studio 14 2015 Update 2 and above (testing with VS 16 2019 Preview)

**Steps To Reproduce**
With Visual Studio 2015 Update 2 or above installed, along with using the native tools x64 command prompt, perform the normal steps, for example:
```
cmake  -G "Visual Studio 16 2019" -A x64 -DBOOST_ROOT="C:\Development\cpp\boost_1_65_1" -DOPENSSL_ROOT_DIR="c:\Development\cpp\openssl-1.1.0f-vs2017" -DBoost_USE_STATIC_LIBS=ON -DOPENSSL_CONF_SOURCE="C:\Development\cpp\openssl-1.1.0f-vs2017\ssl\openssl.cnf" -DCURL_INCLUDE_DIR="C:\Program Files\CURL\include" -DCURL_LIBRARY="C:\Program Files\CURL\lib\libcurl_imp.lib" -DCURL_STATICLIB=ON
cmake --build . --target install --config Release -j 3
```
**Console Logs**
```
6>C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.20.27404\include\functional(1192): error C2027: use of undefined type 'std::_Get_function_impl<_Fty>'
6>        with
6>        [
6>            _Fty=const std::string &
6>        ]
6>C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.20.27404\include\functional(1192): note: see declaration of 'std::_Get_function_impl<_Fty>'
6>        with
6>        [
6>            _Fty=const std::string &
6>        ]
6>C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.20.27404\include\type_traits(1447): note: see reference to class template instantiation 'std::function<const std::string &>' being compiled
6>C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.20.27404\include\type_traits(1556): note: see reference to alias template instantiation 'std::_Is_invocable_r_<graphene::app::order_book,std::function<R (const std::string &,unsigned int)>&,const std::function<const std::string &>&,unsigned int>' being compiled
6>        with
6>        [
6>            R=graphene::app::order_book
6>        ]
6>C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.20.27404\include\functional(1044): note: see reference to class template instantiation 'std::_Is_invocable_r<_Ret,_Fx &,const std::function<const std::string &> &,unsigned int>' being compiled
6>        with
6>        [
6>            _Ret=graphene::app::order_book,
6>            _Fx=std::function<graphene::app::order_book (const std::string &,unsigned int)>
6>        ]
6>C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.20.27404\include\functional(1044): note: see reference to variable template 'const bool conjunction_v<std::negation<std::is_same<std::function<graphene::app::order_book __cdecl(std::basic_string<char,std::char_traits<char>,std::allocator<char> > const &,unsigned int)>,std::function<graphene::app::order_book __cdecl(std::function<std::basic_string<char,std::char_traits<char>,std::allocator<char> > const &> const &,unsigned int)> > >,std::_Is_invocable_r<graphene::app::order_book,std::function<graphene::app::order_book __cdecl(std::basic_string<char,std::char_traits<char>,std::allocator<char> > const &,unsigned int)> &,std::function<std::basic_string<char,std::char_traits<char>,std::allocator<char> > const &> const &,unsigned int> >' being compiled
6>C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.20.27404\include\functional(1207): note: see reference to alias template instantiation 'std::_Func_class<_Ret,const std::function<const std::string &> &,unsigned int>::_Enable_if_callable_t<_Fx&,std::function<R (const std::function<const std::string &> &,unsigned int)>>' being compiled
6>        with
6>        [
6>            _Ret=graphene::app::order_book,
6>            _Fx=std::function<graphene::app::order_book (const std::string &,unsigned int)>,
6>            R=graphene::app::order_book
6>        ]
6>E:\Development\cpp\bitshares-core-2019\libraries\fc\include\fc/rpc/api_connection.hpp(449): note: see reference to function template instantiation 'R fc::generic_api::call_generic<graphene::app::order_book,const std::string&,const std::string&,unsigned int>(const std::function<graphene::app::order_book (const std::string &,const std::string &,unsigned int)> &,std::_Vector_const_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>,std::_Vector_const_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>,uint32_t)' being compiled
6>        with
6>        [
6>            R=graphene::app::order_book,
6>            _Ty=fc::variant
6>        ]
6>E:\Development\cpp\bitshares-core-2019\libraries\fc\include\fc/rpc/api_connection.hpp(195): note: see reference to function template instantiation 'std::function<fc::variant (const fc::variants &)> fc::generic_api::api_visitor::to_generic<graphene::app::order_book,const std::string&,const std::string&,unsigned int>(const std::function<graphene::app::order_book (const std::string &,const std::string &,unsigned int)> &) const' being compiled
6>E:\Development\cpp\bitshares-core-2019\libraries\app\include\graphene/app/database_api.hpp(868): note: see reference to function template instantiation 'void fc::generic_api::api_visitor::operator ()<graphene::app::order_book,const std::string&,const std::string&,unsigned int>(const char *,std::function<graphene::app::order_book (const std::string &,const std::string &,unsigned int)> &) const' being compiled
```

## CORE TEAM TASK LIST
- [ ] Evaluate `Build Error`
- [ ] Provide build guidance
- [ ] <OR> Create `Bug Report`

--- a/tests/tests/htlc_tests.cpp
+++ b/tests/tests/htlc_tests.cpp
@@ -848,7 +848,7 @@ try {
 
    uint16_t preimage_size = 256;
    std::vector<char> pre_image(256);
-   std::independent_bits_engine<std::default_random_engine, CHAR_BIT, unsigned char> rbe;
+   std::independent_bits_engine<std::default_random_engine, sizeof(unsigned), unsigned int> rbe;
    std::generate(begin(pre_image), end(pre_image), std::ref(rbe));
    graphene::chain::htlc_id_type alice_htlc_id_bob;
    graphene::chain::htlc_id_type alice_htlc_id_carl;

--- a/tests/tests/htlc_tests.cpp
+++ b/tests/tests/htlc_tests.cpp
@@ -58,7 +58,7 @@ BOOST_FIXTURE_TEST_SUITE( htlc_tests, database_fixture )
 
 void generate_random_preimage(uint16_t key_size, std::vector<char>& vec)
 {
-	std::independent_bits_engine<std::default_random_engine, CHAR_BIT, unsigned int> rbe;
+	std::independent_bits_engine<std::default_random_engine, sizeof(unsigned), unsigned int> rbe;
 	std::generate(begin(vec), end(vec), std::ref(rbe));
 	return;
 }

